iio: adc: ad5940: Add all ADC channels and a fixed regulator support

cybertale opened the pull request:
These two commits add support for selecting channels from mux to read and the fixed regulator.

commodo commented on drivers/iio/adc/ad5940.c:
```diff
@@ -11,15 +11,151 @@
+
 	switch (info) {
 	case IIO_CHAN_INFO_RAW:
+		switch (chan->channel) {
```
I am wondering here if we could re-use some of the MUX framework that is in `drivers/mux`
or maybe try to re-use some of the logic in the IIO MUX driver (`drivers/iio/multiplexer`)

I am not 100% sure if this is possible.
But I would investigate whether it is possible.

commodo replied:
nevermind; i was wrong;
the MUX framework is not a good idea here

stefpopa commented on drivers/iio/adc/ad5940.c:
```diff
@@ -11,15 +11,151 @@
+			break;
+		default:
+			return -EINVAL;
+		}
```
These channels seem to be highly configurable. Also, some of them can have different functions. I think that the best approach is to define then in the the devicetree. And at probe, we should verify if the configuration is a valid one. 
I suggest that you use as inspiration the AD7124 ADC which also has configurable channels.
https://github.com/analogdevicesinc/linux/blob/master/drivers/iio/adc/ad7124.c#L433
https://github.com/analogdevicesinc/linux/blob/master/Documentation/devicetree/bindings/iio/adc/adi%2Cad7124.txt
https://www.analog.com/media/en/technical-documentation/data-sheets/ad7124-8.pdf

cybertale replied:
Thanks, I'll look into this.

stefpopa commented on drivers/iio/adc/ad5940.c:
```diff
@@ -33,17 +169,33 @@ static int ad5940_read_raw(struct iio_dev *indio_dev,
 		.channel = idx,						\
 		.info_mask_separate = BIT(IIO_CHAN_INFO_RAW),		\
 		.info_mask_shared_by_type = BIT(IIO_CHAN_INFO_SCALE),	\
+		.extend_name = name,					\
```
This changes the ABI, and therefore needs to be documented.  A preferred way to deal with this situations is to create an attribute which will return the datasheet channel name. Here's an example which uses this approach: https://github.com/analogdevicesinc/linux/blob/master/Documentation/ABI/testing/sysfs-bus-iio-frequency-adf4371#L26

stefpopa commented on drivers/iio/adc/ad5940.c:
```diff
@@ -11,15 +11,151 @@
+	CHAN_TEMP,	/* Internal Temperature Sensor. */
+	CHAN_AVDD_2,	/* Half of the AVDD. */
+	CHAN_VREF_2V5_2,	/* Half of the 2.5v reference voltage. */
+	CHAN_VREF_1V82,
```
The enum elements should have as prefix the module name: `AD5940_CHAN_CE0`

stefpopa commented on drivers/iio/adc/ad5940.c:
```diff
@@ -221,6 +236,20 @@ static int ad5940_probe(struct spi_device *spi)
 
+	st->vref = devm_regulator_get(&spi->dev, "vref");
+	if (IS_ERR(st->vref))
+		return PTR_ERR(st->vref);
```
Let's make this property optional. If the regulator is present in the devicetree, then we know that we should use an external vref, otherwise, we use the internal vref which can be hardcoded. Here's an example: https://elixir.bootlin.com/linux/v5.2-rc6/source/drivers/iio/dac/ad5686.c#L440

stefpopa commented on drivers/iio/adc/ad5940.c:
```diff
@@ -160,10 +166,12 @@ static int ad5940_read_raw(struct iio_dev *indio_dev,
+		if (scale_uv < 0)
+			return scale_uv;
+		*val = scale_uv / 1000;
+		*val2 = AD5940_REALBITS;
```
For now, I would use value 15 with a comment, instead of this define . Later, when we will add buffer support, we will get this info from `chan->scan_type.realbits`

cybertale commented:
### Change log

1. Deleted some left interrupt related stuff.
2. Moved channel definition to devicetree and added some parse code.
3. Moved channel name to iio_chan_spec_ext_info and created a ABI doc for it.
4. Added a devicetree property 'vref-uv' to convey internal reference voltage value. This value will be used when 'vref-supply' is not presented in devicetree, that is when we are using external regulators.
5. Made 'vref-supply' in devicetree optional.
6. Removed the fixed regulator node in rpi-ad5940-overlay.dts.
6. Deleted AD5940_REALBITS macro.
@stefpopa @commodo @dbogdan

nunojsa commented on drivers/iio/adc/ad5940.c:
```diff
@@ -5,14 +5,71 @@
+		return -EINVAL;
+	}
+
+	return 0;
```
This return is not needed...

cybertale replied:
I see, thanks!

nunojsa commented on Documentation/ABI/testing/sysfs-bus-iio-adc-ad5940:
```diff
@@ -0,0 +1,7 @@
+Contact:	songqiang1304521@gmail.com
+Description:
+		Channel name assigned in devicetree, used to tell userspace
+		users whcih channel this is.
```
typo in "whcih"

stefpopa commented on drivers/iio/adc/ad5940.c:
```diff
@@ -5,14 +5,71 @@
+	case AD5940_CHANNEL_NAME:
+		return sprintf(buf, "%s\n",
+			st->channel_config[chan->address].channel_name);
+		break;
```
`break `statement is never reached

stefpopa commented on Documentation/devicetree/bindings/iio/adc/adi,ad5940.yaml:
```diff
@@ -31,6 +36,7 @@ properties:
 required:
   - compatible
   - reg
+  - vref-uv
```
this property should be optional, in case it is not specified, we shall use the internal vref

stefpopa commented on drivers/iio/adc/ad5940.c:
```diff
@@ -33,23 +90,114 @@ static int ad5940_read_raw(struct iio_dev *indio_dev,
+	for (i = 0; i < ARRAY_SIZE(channel_p); i++) {
+		if (ain[0] == channel_p[i]) {
+			ret = 0;
+			break;
```
I think that we can use` bsearch()`, here's an example: https://elixir.bootlin.com/linux/v5.2-rc7/source/drivers/iio/dac/ad5758.c#L710

stefpopa commented on drivers/iio/adc/ad5940.c:
```diff
@@ -217,6 +235,41 @@ static int ad5940_probe(struct spi_device *spi)
+	if (ret) {
+		dev_err(&spi->dev, "Failed to get vref-uv from device tree.\n");
+		return ret;
+	}
```
It's better to use `device_property_read_u32()`; also, if this property is not present in the devicetree, then we should use the internal reference, therefore we should not return an error.

stefpopa commented on drivers/iio/adc/ad5940.c:
```diff
@@ -217,6 +235,41 @@ static int ad5940_probe(struct spi_device *spi)
+		ret = devm_add_action_or_reset(&spi->dev,
+				ad5940_regulator_disable, st);
+		if (ret)
+			goto err_disable_reg;
```
If we move `devm_add_action_or_reset()` right after the regulator was enabled, then there is no need for a goto statement. The `devm*` will take care of `regulator_disable()`

stefpopa commented on drivers/iio/adc/ad5940.c:
```diff
@@ -232,6 +285,11 @@ static int ad5940_probe(struct spi_device *spi)
+err_disable_reg:
+	if (!IS_ERR(st->vref))
+		regulator_disable(st->vref);
+
```
As mentioned in the above comment, the whole point of using device managed functions is to get rid of this whole disable, unprepare, unregister stuff :smiley:

cybertale commented:
> Looks pretty good overall, just a few minor remarks

Thanks, I'll fix the remaining problems tomorrow. :smile:

cybertale commented:
# Change Log

1. Fixed the 'whcih' typo.
2. Made vref-uv to be optional and added some description.
3. Deleted unnecessary 'break' and 'return'.
4. Using bsearch to search for channel indexes.
5. Added 'channel' to the error output on line 185.
6. Moved devm_add_action_or_reset right after regulator_enable and removed the 'goto' statements.
7. Moved operation of retrieving 'vref-uv' in the 'devm_regulator_get_optional' block, which makes it to be executed only when no external regulators are presented.
8. Using device_property_read_u32 to read 'vref-uv'.

stefpopa commented on Documentation/devicetree/bindings/iio/adc/adi,ad5940.yaml:
```diff
@@ -107,6 +113,7 @@ examples:
       compatible = "adi,ad5940";
       reg = <0>;
       spi-max-frequency = <16000000>;
+      vref-uv = <1820000>;
```
we do not need a custom property for this, just use the `fixed regulator`, therefore this property should be called vref-supply. Here's an overlay example which uses the fixed regulator: https://github.com/analogdevicesinc/linux/blob/rpi-4.14.y/arch/arm/boot/dts/overlays/rpi-ad7768-overlay.dts

stefpopa commented on drivers/iio/adc/ad5940.c:
```diff
@@ -216,6 +234,41 @@ static int ad5940_probe(struct spi_device *spi)
+					       &internal_vref_uv);
+		if (ret) {
+			dev_err(&spi->dev, "Failed to get vref-uv from device tree.\n");
+			return ret;
```
so, as mentioned before there is no need for a custom property

stefpopa commented on drivers/iio/adc/ad5940.c:
```diff
@@ -216,6 +234,41 @@ static int ad5940_probe(struct spi_device *spi)
+	if (vref_uv)
+		st->vref_mv = vref_uv / 1000;
+	else
+		st->vref_mv = internal_vref_uv / 1000;
```
the problem here is that if the vref regulator is not present in the devicetree, then internal_vref_uv variable is undefined. We need a hard coded value for the internal reference: `unsigned int int_vref_mv = 1820; `

cybertale replied:
So if I'm understanding this right, it means that we can present the internal reference with the fixed regulator or hard coded value. We search for the 'vref-supply', which could be the internal reference presented as a fixed regulator or any external one, and if we cannot find it, we use the hard coded value instead, right?

stefpopa replied:
That's almost correct :smiley: 
So, the ad5940 ADC has a 1.82 V internal reference, however an external reference source can also be connected. This second option is not the default one and therefore should be optional.

How we usually deal with this cases is that we consider that if a fixed regulator property is present in the devicetree, that means that an external reference needs to be used. This is how you use the fixed regulator in the devicetree overlay: https://github.com/analogdevicesinc/linux/blob/rpi-4.14.y/arch/arm/boot/dts/overlays/rpi-ad5686-overlay.dts#L10-L15. And the regulator will be called `vcc-supply`: https://github.com/analogdevicesinc/linux/blob/rpi-4.14.y/arch/arm/boot/dts/overlays/rpi-ad5686-overlay.dts#L32

Next, in the code we need to call `devm_regulator_get_optional()` and check if vcc is present in the devicetree: https://github.com/analogdevicesinc/linux/blob/master/drivers/iio/dac/ad5686.c#L441

If this property is not present, then we use the internal reference: https://github.com/analogdevicesinc/linux/blob/master/drivers/iio/dac/ad5686.c#L459
The internal reference can simply be hardcoded: https://github.com/analogdevicesinc/linux/blob/master/drivers/iio/dac/ad5686.c#L270

cybertale commented:
# Change Log

1. Removed the 'vref-uv' property from devicetree and added a fixed regulator to it.
2. Added a const value to supply a hard coded internal voltage reference value if none vref-supply is presented in devicetree.

stefpopa commented on Documentation/devicetree/bindings/iio/adc/adi,ad5940.yaml:
```diff
@@ -107,6 +112,7 @@ examples:
       compatible = "adi,ad5940";
       reg = <0>;
       spi-max-frequency = <16000000>;
+      vref-uv = <1820000>;
```
you forgot to remove `vref-uv` property

stefpopa commented on drivers/iio/adc/ad5940.c:
```diff
@@ -216,6 +234,33 @@ static int ad5940_probe(struct spi_device *spi)
+	if (vref_uv)
+		st->vref_mv = vref_uv / 1000;
+	else
+		st->vref_mv = int_vref_mv;
```
Maybe set it directly: 
```
else
	st->vref_mv = 1820;
```

stefpopa commented:
I will merge this later today if there are no objections.

cybertale commented:
> I will merge this later today if there are no objections.

Thanks!

commodo commented on Documentation/devicetree/bindings/iio/adc/adi,ad5940.yaml:
```diff
@@ -22,14 +22,190 @@ properties:
+      channel@0 {
+        reg = <0>;
+        diff-channels = <AD5940_ADC_INPUTP_VCE0
+              AD5940_ADC_INPUTN_VBIAS0>;
```
sorry to be late at the review party :)
can this be indented neatly like in the ` arch/arm/boot/dts/overlays/rpi-ad5940-overlay.dts ` file?
it is in the docs after all

cybertale replied:
Sure, that's better

commodo commented on drivers/iio/adc/ad5940.c:
```diff
@@ -73,16 +77,21 @@ static const struct iio_chan_spec ad5940_channel_template = {
+		scale_uv = regulator_get_voltage(st->vref);
+		if (scale_uv < 0)
+			return scale_uv;
+		*val = st->vref_mv;
```
i am a bit confused here;
maybe i missed a discussion point
shouldn't we use the value from `scale_uv` here?

maybe we could wrap the logic into a function an re-use it?
something like?
```
int ad5940_get_ref_voltage(struct ad5940_state *st)
{
        int ret;
        if (!st->vref)
              return 1820;

	ret = regulator_get_voltage(st->vref);
        if (ret < 0)
               return ret;
	}

         return (ret / 1000)
```

cybertale replied:
Indeed, we do not need to get voltage here again since regulators and voltage are determined in the probe function, so here we should just return the scale_mv without any condition.

commodo replied:
that also works for me

cybertale commented:
# Change Log v6
1. Made indent of diff-channels in the yaml files better.
2. Removed out dated code in returning scale values.
